builtin/gc.c

```diff
@@ -1345,6 +1348,131 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 	return 0;
 }
 
+static void link_or_copy_or_die(const char *src, const char *dst)
+{
+	if (!link(src, dst))
+		return;
+
+	warning_errno(_("failed to link '%s' to '%s'... trying copy..."), src,
+		      dst);
+
+	if (!copy_file(dst, src, 0644))
+		return;
+
+	die_errno(_("failed to copy '%s' to '%s'"), src, dst);
+}
+
+static void migrate_pack(const char *srcdir, const char *dstdir,
+			 const char *pack_filename)
+{
+	struct stat st;
+	int has_keep, has_rev, has_idx;
+	char *basename, *pack_src, *keep_src, *rev_src, *idx_src,
+			*pack_dst, *keep_dst, *rev_dst, *idx_dst;
+
+	trace2_region_enter("maintenance", "migrate_pack", the_repository);
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
+	pack_src = xstrfmt("%s/%s", srcdir, pack_filename);
+	pack_dst = xstrfmt("%s/%s", dstdir, pack_filename);
+	keep_src = xstrfmt("%s/%s.keep", srcdir, basename);
+	keep_dst = xstrfmt("%s/%s.keep", dstdir, basename);
+	rev_src = xstrfmt("%s/%s.rev", srcdir, basename);
+	rev_dst = xstrfmt("%s/%s.rev", dstdir, basename);
+	idx_src = xstrfmt("%s/%s.idx", srcdir, basename);
+	idx_dst = xstrfmt("%s/%s.idx", dstdir, basename);
+
+	has_keep = !stat(keep_src, &st);
+	has_rev = !stat(rev_src, &st);
+	has_idx = !stat(idx_src, &st);
+
+	/* A pack without an index file is not yet ready to be migrated. */
+	if (!has_idx)
+		goto cleanup;
+
+	/*
+	 * Hard link (or copy if that fails) all but the index file so that
+	 * other Git processes don't attempt to use the pack file from the new
+	 * location yet.
+	 */
+	link_or_copy_or_die(pack_src, pack_dst);
+	if (has_keep)
+		link_or_copy_or_die(keep_src, keep_dst);
+	if (has_rev)
+		link_or_copy_or_die(rev_src, rev_dst);
+
+	/*
+	 * Move the index file atomically now that the other files can be found
+	 * at the destination.
+	 */
+	if (rename(idx_src, idx_dst))
+		die_errno(_("failed to move '%s' to '%s'"), idx_src, idx_dst);
+
+	/*
+	 * Now the pack and all associated files exist at the destination we can
+	 * now clean up the files in the source directory.
+	 */
+	if (unlink(pack_src))
+		warning_errno(_("failed to delete '%s'"), pack_src);
+	if (has_keep && unlink(keep_src))
+		warning_errno(_("failed to delete '%s'"), keep_src);
+	if (has_rev & unlink(rev_src))
+		warning_errno(_("failed to delete '%s'"), rev_src);
+
+cleanup:
+	free(idx_src);
+	free(idx_dst);
+	free(rev_src);
+	free(rev_dst);
+	free(keep_src);
+	free(keep_dst);
+	free(pack_src);
+	free(pack_dst);
+	free(basename);
+
```

Comment:
One way to avoid some of this repeated logic would be to have an array of pairs (from/src) where if the first one is non-NULL then we act upon it. The order of the array can be constructed to follow the order we want. It can then be expanded to include a new file type (e.g. `.promisor`, `.mtimes` though they shouldn't happen here) with minimal code duplication.